File: 382-linked-list-random-node/382-linked-list-random-node.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def __init__(self, head: Optional[ListNode]):
        self.list = head
        
    def getRandom(self) -> int:
        head = self.list
        r = head.val
        n = 1
        head = head.next
        
        while head != None:
            if(random.randint(0, n) == 0):
                r = head.val
            head = head.next
            n += 1
        return r


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()

File: 382-linked-list-random-node/382-linked-list-random-node.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* list;
    
public:
    Solution(ListNode* head) {
        list = head;
    }
    
    int getRandom() {
        auto head = list;
        int r = head -> val;
        int n = 1;
        head = head -> next;
        
        while(head) {
            if(rand() % (n + 1) == 0)
                r = head -> val;
            head = head -> next;
            ++n;
        }
        return r;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(head);
 * int param_1 = obj->getRandom();
 */